Implement MsrList and Msrs as FamStructWrappers

acatangiu opened the pull request:
# Description of changes

`kvm_msr_list` and `kvm_msrs` each contain a flexible array member and were forcing users of `kvm-ioctls` crate to work with `unsafe` code.
Use the `kvm-bindings` `CpuId`, `MsrList` and `Msrs` safe wrappers over them and change our ioctl functions to use these safe wrappers.

jiangliu commented:
Could we implement this in kvm-bindings crate? It should be simpler and natural to implement mechanism to access syscall related structs within the kvm-bindings crate.

andreeaflorescu commented:
@jiangliu I don't think that implementing these traits in the kvm-bindings crate is a good idea at the moment. There are a few reasons for that:

1. kvm-bindings still separates clearly the autogenerated code from the one that is not autogenerated. I would prefer to keep it this way for now.
2. we are in discussions about merging/collaborating with ketuvim (https://github.com/rust-vmm/community/issues/62). Probably it will become more clear after the rust-vmm meetup what is the plan there.

While I believe that it's not the best thing now to move this functionality in kvm-bindings, we can discuss about it again once the rust-vmm components and their separation are more obvious.

What do you say?

jiangliu commented:
> @jiangliu I don't think that implementing these traits in the kvm-bindings crate is a good idea at the moment. There are a few reasons for that:
> 
> 1. kvm-bindings still separates clearly the autogenerated code from the one that is not autogenerated. I would prefer to keep it this way for now.
When enabling serde for kvm-bindings, I ran into the same issue to add non-auto-generated code to the kvm-bindings crate. In some sense, we are enhancing the bindgen tool to better support FAM structs, so 
propose to implement in the kvm-binding crate.
It's great to collaborate with the ketuvim project, which may help to solve all these issues.

> 2. we are in discussions about merging/collaborating with ketuvim ([rust-vmm/community#62](https://github.com/rust-vmm/community/issues/62)). Probably it will become more clear after the rust-vmm meetup what is the plan there.
> 
> While I believe that it's not the best thing now to move this functionality in kvm-bindings, we can discuss about it again once the rust-vmm components and their separation are more obvious.
> 
> What do you say?

acatangiu commented:
After some thought, I agree that the FamWrappers over the kvm FAM structures belong in the same crate as the structs themselves (kvm-bindings). Doing that also simplifies the implementation since we can directly implement the external FamStructWrapper trait on a local struct.

I've opened https://github.com/rust-vmm/kvm-bindings/pull/8 that exemplifies how that'd look.
The PR adds a new module in that crate that doesn't touch the generated files, it just imports the structs from there, so that's nice. The downside is that `kvm-bindings` would depend on `vmm-sys-utils`.

acatangiu committed:
add .idea to .gitignore

Signed-off-by: Adrian Catangiu <acatan@amazon.com>

acatangiu committed:
use safe fam-wrapper CpuId from kvm-bindings

Use safe fam-wrapper over kvm_cpuid2 exported by the
kvm-bindings crate instead of implementing our own.

Coverage decreases by 0.1% because of the removed code.

Signed-off-by: Adrian Catangiu <acatan@amazon.com>

acatangiu committed:
use safe wrappers Msrs and MsrList from kvm-bindings

Refactor the unsafe code concerning kvm_msr_list and kvm_msrs,
to work with the safe fam-wrappers MsrList and Msrs.

Signed-off-by: Adrian Catangiu <acatan@amazon.com>

acatangiu commented:
The PR is updated to use the `kvm-bindings` safe wrappers. Please review when you get the chance.
